flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1115,6 +1117,15 @@ mlir::Value IntrinsicLibrary::genDim(mlir::Type resultType,
   return builder.create<mlir::SelectOp>(loc, cmp, diff, zero);
 }
 
+// DPROD
+mlir::Value IntrinsicLibrary::genDprod(mlir::Type resultType,
+				       llvm::ArrayRef<mlir::Value> args) {
+  assert(args.size() == 2);
+  assert(fir::isa_real(resultType) && (resultType.getKind() == 8) && "Result must be double precision in DPROD");
```

Comment:
Are you sure real(8) is guaranteed to be double precision ? That looks OK to me with flang current configuration, but I am not entirely sure this is guaranteed (maybe it is, I did not double check).

Reply:
Would `(resultType.getKind() == mlir::StandardTypes::F64)` work better instead?  I came across that, but I also was not sure how to guarantee double precision and I'm just used to seeing `REAL(8)` used for that purpose.  It looks like `mlir::StandardTypes::F64` gets converted to a double when MLIR is lowered to LLVM (line 199-200 in `mlir/lib/Conversion/StandardToLLVM/StandardToLLVM.cpp`).

Reply:
I suspect that semantics should be verifying that constraint/requirement.  Because we have two type systems in play here (FIR and MLIR), we might want to make a generalized test at some point.

But, in general, one shouldn't rely on hard-coded KIND values but rather use the KindMapping code.

Reply:
I looked a bit more in the standard in the front-end, and I agree with Eric that **we should not assert on what floating point type `resultType` is**. I think we can neither be sure that double complex ends up `mlir::StandardTypes::F64` or KIND 8 here.

**Regarding the KIND of DOUBLE PRECISION:**

- 7.4.3.2 point 4 `The keyword DOUBLE PRECISION is an alternative specifier for one kind of real type.` without specifying which kind.
So the compiler is free to use something else as `8` for double precision, f18 uses `8` by deafult, but this can be change at (Fortran) compile time in f18, see `IntrinsicTypeDefaultKinds` class:
https://github.com/flang-compiler/f18-llvm-project/blob/1f8790050b0e99e7b46cc69518aa84f46f50738e/flang/include/flang/Common/default-kinds.h#L23-L26
f18 option `-fdefault-real-8` will actually set double precision to be 16 and not 8 (we cannot really play with that in bbc yet).

**Regarding the underlying float type of DOUBLE PRECISION:**
Fortran does not require double precision to be a 64bit IEEE float. It only says:
- 7.4.3.2 point 5: `The decimal precision of double precision real shall be at least 10, and its decimal exponent range shall be at
least 37.`
- 19.5.3.2 point 2 (2) : `a nonpointer scalar object that is double precision real or default complex occupies two contiguous
numeric storage units`

So it can be a 64bit IEEE float, but may not be. Still with the `-fdefault-real-8`, we would end up with REAL(16) that is currently implemented with 128bit IEEE float, so your assert would fire here.

**Why can we trust resultType is DOUBLE PRECISION ?**
`resultType` of intrinsic is indirectly coming from semantics analysis of intrinsic, it is not a user input, so we can trust it here. The result type of intrinsics are defined here:
https://github.com/flang-compiler/f18-llvm-project/blob/1f8790050b0e99e7b46cc69518aa84f46f50738e/flang/lib/Evaluate/intrinsics.cpp#L358

**Why do we still hard code the mlir/fir types in LIT tests then ?**

Based on the above point, you could argue that the LIT tests should also not hard code `f32`, `f64` .... for a Fortran program using REAL/DOUBLE PRECISION. However, it turns out to be really hard to avoid that in the test, and also we are in control of the test options: i.e we now we are not passing `-fdefault-real-8` option. That only means we should add at some point a Dprod test with `f64`, `fir.real<16>` (no f128 in mlir) with the option `-fdefault-real-8` (not yet possible since bbc does not handle that).

Reply:
Thank you both for taking the time to explain this to me.  @jeanPerier , I really appreciate the level of detail.  This has been a very helpful learning experience.